test/studies/adventOfCode/2022/day12/jeremiah/day12a-par.chpl

```diff
@@ -0,0 +1,92 @@
+param a = b"a".toByte(),
+      S = b"S".toByte(),
+      E = b"E".toByte();
+
+const (elevations, start, end) = readElevations();
+const validMoves = findValidMoves(elevations);
+writeln(findShortestPath(start, end, validMoves));
+
+proc readElevations() {
+    use IO;
+
+    const elevBytes = stdin.readAll(bytes),
+          w = elevBytes.find(b"\n"),
+          h = elevBytes.count(b"\n");
+
+    var elevs : [0..<h, 0..<w] int(8),
+        start, end = (0, 0);
+
+    for (row, i) in zip(elevBytes.split(b"\n"), 0..) {
+        foreach (byte, j) in zip(row, 0..) {
+            select byte {
+                when S do start = (i, j);
+                when E do end = (i, j);
+                otherwise do elevs[i, j] = (byte - a) : int(8);
+            }
+        }
+    }
+
+    elevs[start] = 0;
+    elevs[end] = 25;
+
+    return (elevs, start, end);
+}
+
+proc findValidMoves(elevs: [?d] int(8)) : [d] uint(8) {
+    var moves: [d] uint(8);
+    var _elevs : [d.expand(1)] int(8) = 28;
+    _elevs[d] = elevs;
+
+    forall (i, j) in d do
+       for (m, di, dj) in ((1, 0, 1), (2, 0, -1), (4, 1, 0), (8, -1, 0)) do
+            moves[i, j] +=
+                (if elevs[i, j] - _elevs[i+di, j+dj] >= -1 then m else 0) : uint(8);
+
+    return moves;
+}
+
+iter openMoves(m: uint(8)) {
+    if m & 1 then yield (0, 1);
+    if m & 2 then yield (0, -1);
+    if m & 4 then yield (1, 0);
+    if m & 8 then yield (-1, 0);
+}
```

Comment:
I'm not snapping to what all the bitmask / 3-tuple logic is doing in these two routines, and feeling nervous that this program is ~50% longer than my version.  Can you say more about what's happening here and why?

Reply:
Rather than looking at the elevations on the fly in order to determine which directions can be explored next, I decided to create an array ahead of time that holds the valid directions at each grid point. The array is then queried during "exploration" which makes the explore function look a lot simpler IMO (around line 88 specifically, there is no elevation-related logic).

I'm storing the valid directions at each grid-point as a single number with some bitmask logic where:
* 1 => right
* 2 => left
* 4 => up
* 8 => down

Not sure if it's better or worse, but I could rewrite:
``` chapel
forall (i, j) in d do
       for (m, di, dj) in ((1, 0, 1), (2, 0, -1), (4, 1, 0), (8, -1, 0)) do
            moves[i, j] +=
                (if elevs[i, j] - _elevs[i+di, j+dj] >= -1 then m else 0) : uint(8);
```
as:
```chapel
forall idx in d do
      for (m, dIdx) in ((1, (0, 1)), (2, (0, -1)), (4, (1, 0)), (8, (-1, 0)))
            moves[idx] +=
                (if elevs[idx] - _elevs[idx + dIdx] >= -1 then m else 0) : uint(8);

```
which might make it more clear that `m` is a mask, and the adjacent two-tuples are the corresponding directions?

---
> feeling nervous that this program is ~50% longer than my version

The bitmask stuff does add a bit of length, but I think most of it is coming from the `readElevations` proc which is much longer than your IO code. I can definitely trim that down though.

Reply:
@jeremiah-corrado : Even with the explanation, the bitmask stuff still feels overwrought to me—specifically the magic numbers of 1, 2, 4, 8 and the use of the 3-tuples (or 2-deep tuples) for a 2D problem.  If we end up keeping it, I'd at least strongly suggest defining a concrete enum (or series of params) for the bitmasks so that they have symbolic names.

That said, when you spoke of moving elevation stuff into the iterator, I was imagining something like the following, which I find very attractive:

```chapel
// an iterator that yields all possible dirs to explore
iter allDirs() {
  yield (0,1);  // or name these using consts: south, east, west, north
  yield (1,0);
  yield (-1,0);
  yield (0,-1);
}

// an iterator that yields new positions we should explore from the current one
iter newPositions(currentPos) {
  for dir in allDirs() {
    const newPos = pos + dir;
    if elevation[newPos] <= elevation[currentPos+1]+1 {   // I may not have the comparison correct here...
      yield newPos;
}
```

I'd also consider pushing the shortest-path-to-this-pos-so-far check into this second iterator to avoid spawning recursions/tasks that we know aren't worthwhile (and maybe the "exceeded shortest path to endpoint" check as well for similar reasons?).  Beyond the savings in tasks/time, this would also make the explore routine even simpler.

Since the elevation array is constant / non-atomic, it should be cheap to access relative to the atomic variables if that was part of your concern in pre-computing legal moves.  Also, note that my reaction to the bitmask approach is less about using the approach itself "in real life" and more about keeping the article focused on simpler codes that illustrate unique Chapel features and their uses than approaches that are more complex/heroic/clever than need be.  Put another way, I think understanding multitasking, sync statements, and atomics is a pretty steep learning curve for most programmers to begin with—so keeping the computational part as straightforward as possible seems to the article's benefit.

Reply:
So I went down the rabbit hole of coding up what I proposed above to see how much it helped.  The quick takeaways are:
* I believe it cut down on the number of tasks spawned / recursive calls made significantly relative to my previous versions—less sure about relative to yours, though I'd expect a modest improvement due to moving more short-circuiting logic into the iterator and out of Explore()
* I believe I found a bug in `begin`s w.r.t. tuples that is making me nervous about using them here.  I've asked Vass to take a look.  I've got a workaround but I don't like the thought of having to rationalize it in a blog.  Have you tried stress-testing your solution over 1000's of runs to see if you get any bad results?
* Meanwhile, I found that if I switched from for + begin to coforall, I used significantly fewer tasks on average.  I reached for begin originally because it felt like that kind of asynchrony ("You go your way and I'll go mine and we never have to join up again").  But something about the coforall—maybe the order in which tasks are scheduled and considered to do more unique work sooner?—seems to give significant savings.

As a result of this, my current favorite code is in day12-par.chpl at https://github.com/chapel-lang/chapel/pull/21232/files  The version that seems to show the bug and support comparisons between different techniques is in day12-par-multiversion.chpl where versions 0 (begin with workaround) and 2 (coforall) work and version 1 (begin without workaround) doesn't for reasons I can't fathom other than blaming a bug so far...

Reply:
> Since the elevation array is constant / non-atomic, it should be cheap to access relative to the atomic variables if that was part of your concern in pre-computing legal moves

fwiw, I was mainly aiming for a solution that followed the flow of the problem description, where they showed a directions array as an intermediate step to finding the shortest path:
```
v..v<<<<
>v.vv<<^
.>vv>E^^
..v>>>^^
..>>>>>^
```
And I also thought it was a cleaner way of handling the edges (the `if grid.conatains(next)` check isn't necessary in my code during the exploration phase).

>  I think understanding multitasking, sync statements, and atomics is a pretty steep learning curve for most programmers to begin with—so keeping the computational part as straightforward as possible seems to the article's benefit.

Yeah, that's a good point. I'll go with a solution like what you've linked above. Specifically, it sounds like the `coforall` version is best due to the bug you mentioned. 

In that version, is the `sync { ... }` still necessary since it won't be using `begin` statements?

> Have you tried stress-testing your solution over 1000's of runs to see if you get any bad results?

I haven't, but I did notice that it was necessary to precompute `next` before spawning a new task. I.e., this worked:
```chapel
const next = move + pos;
begin explore(next, end, moves, minDistances, shortestSoFar, pathLen + 1);
```
but this didn't:
```chapel
begin explore(move + pos, end, moves, minDistances, shortestSoFar, pathLen + 1);
```
however, I figured that was just a quirk of begin statements.

I can definitely run a stress test on my working version if you think that'd be helpful for tracking down this bug. Let me know.

Reply:
> I was mainly aiming for a solution that followed the flow of the problem description, where they showed a directions array as an intermediate step to finding the shortest path:

I think that possibility occurred to me late last night...  I think it's a little different in that their map only shows the directions they explored, not all possible directions one could go from a place, though, is that right?  Anyway, thanks for taking the suggestion to simplify that logic—maybe I'm just biased because it's the way I wrote it at first, but it seems much more straightforward to me, and akin to the way one would learn to walk a tree recursively, with aggressive pruning.

> In that version, is the sync { ... } still necessary since it won't be using begin statements?

You're guessing right that it's not.  Oh, but I didn't notice that you still have it on your branch.

> however, I figured that was just a quirk of begin statements.

That sounds like it's probably the same bug I ran into last night—I don't think it should matter.  Vass seemed to validate that it was a bug based on his understanding of what should happen and the spec which was both reassuring (I'm not crazy / this language isn't crazy subtle) and worrying (in that something that feels pretty fundamental isn't acting as it's meant to).  I haven't gone back to see whether this is something that broke in recent releases yet...

I did do a 1000-trial stress test on the version you pointed us to this morning and it seemed fine, which is not as surprising since it's using the coforall.

Reply:
> > In that version, is the sync { ... } still necessary since it won't be using begin statements?

> You're guessing right that it's not. Oh, but I didn't notice that you still have it on your branch.

Makes sense, I'll remove the `sync` from the final version.

> I did do a 1000-trial stress test on the version you pointed us to this morning and it seemed fine

Ok great, thanks for checking that!